### DISCUSSION TITLE
Add wait() method to Job Batching

### DISCUSSION THREAD

**Discussion by handhikadj** [laravel/framework#52528] [Ideas]
currently there's no way to wait for a job batch to finish. adding `wait()` method to `Batch` would be awesome. something like this:

```php
public function aMethodInController()
{
  $batch = Bus::batch([]);
  
  foreach (range(1, 20) as $numb) {
    $batch->add(new MyQueue($numb));
  }

  $batch
    ->dispatch()
    ->wait();

  return view('job_batch_completed_view');
}  
```

I think this will open parallel processing capability to the framework

**Comment by henzeb** [comment]
I don't think parallel processing should go in a controller ... If you must, you better use the sync driver without batch.

**Comment by ejunker** [comment]
I implemented it using polling like this:

```php
  protected function waitForBatch(Batch $batch, \Closure $callback, $pollSeconds = 5): void
  {
      while (! $batch->finished()) {
          $batch = $batch->fresh();
          $callback($batch);
          sleep($pollSeconds);
      }
  }
```

Then when calling it you can do:

```php
$this->waitForBatch($batch, function (Batch $batch) use (&$progress) {
    $currentProgress = $batch->progress();
    if ($currentProgress !== $progress) {
        $this->info('['.now()->toDateTimeString().'] '.$batch->processedJobs().'/'.$batch->totalJobs.' '.$batch->progress().'%');
        $progress = $currentProgress;
    }
});
```

**Reply by jordanade** [reply]
@ejunker hey this is pretty good, but when jobs in the batch fail they cause the code above to hang forever. Any idea on a fix for that?

**Reply by ejunker** [reply]
@jordanade you might have to call `allowFailures()` on your batch or if you don't want to do that you would also have to check if the batch is `canceled()` in the while loop condition

**Reply by jordanade** [reply]
Hmm, I did both but still had the issue. My jobs fail with Illuminate\Queue\TimeoutExceededException.

What finally worked was modifying the function as so:
```    
protected function waitForBatch(Batch $batch, \Closure $callback, $pollSeconds = 2): void
{
    while ($batch->pendingJobs - $batch->failedJobs) {
        $batch = $batch->fresh();
        $callback($batch);
        sleep($pollSeconds);
    }
}
```
